### PR TITLE
fix(assistant-stream): prioritize backend tool results

### LINKED ISSUE #5102: assistant-stream: known-divergent argsText close becomes an auto-submitted error tool result
## Summary

The adapter-layer causes of args divergence keep getting fixed one dimension at a time (#3458 key order, #3780 tuple path, #5099 updates events, #5080 argless first chunks), but the tracker layer still converts any remaining divergence into a hard failure: an error tool result that hosts auto-submit. This issue tracks hardening that conversion.

## Mechanism (verified on main with a unit repro during the #5098 verification)

1. A snapshot's `argsText` diverges from the tracked prefix while the tracked prefix is not valid JSON. `ToolInvocationTracker._processArgsText` takes the A.2 path and keeps the stale prefix (`packages/core/src/runtimes/tool-invocations/ToolInvocationTracker.ts`, A.2 branch).
2. The close block then computes `_shouldCloseArgsStream` from the snapshot's complete `argsText` and closes the controller that still holds the stale prefix.
3. `ToolExecutionStream` runs strict `sjson.parse` on the stale prefix, throws `Function parameter parsing failed`, and its error handler wraps that into `ToolResponse({ isError: true })` (`packages/assistant-stream/src/core/tool/ToolExecutionStream.ts`).
4. External-store hosts receive it through `onAddToolResult` and auto-submit; on `useLangGraphRuntime` this resumes the graph and can abandon a pending interrupt (see #5098 for the full production consequence).

The tracker cannot adopt the divergent snapshot today: the `argsText` stream is append-only (`ToolCallStreamController.argsText` exposes append/close only, no replace), so appending cannot reconcile divergent text; the only tracker-level options are closing with a broken prefix (parse error) or never closing (hang).

`EDGE_CASES.md` "Known limitations" already documents the sibling shape (A.2 + A.5 in the same snapshot) and points at `ToolCallStreamControllerImpl.setResponse` enqueueing the result before closing `argsText` as the upstream fix.

## Remaining triggers after #5099 / #5080

- any adapter that re-stringifies final args instead of carrying the streamed text
- mechanism-level example in react-langgraph: `LangGraphMessageAccumulator.reconcileMessages` / `replaceMessages` bypass `appendMessage`, so a values-mode final reconcile can reintroduce a compact re-serialization under extreme render batching (mechanism verified, not reproduced end to end)

## Directions

1. Fix the `setResponse` ordering per the EDGE_CASES.md note, so a backend result is not beaten by the parse-error result.
2. When a close follows a known-divergent snapshot whose `argsText` is complete valid JSON, let the tracker hand that snapshot text to the execution path instead of the stale stream prefix (requires a small assistant-stream API addition, since the stream itself never sees snapshots).
3. At minimum, stop auto-submitting an error result that was produced by this internal bookkeeping failure rather than by the tool or the model.

### PR DESCRIPTION
**Pull request by Solaris-star**
## Summary

- emit backend tool responses before closing the argument text stream
- prevent a stale argument parse failure from winning the response race after `argsText` divergence
- add an assistant-stream regression test for the response ordering and update the documented tracker edge case

## Root cause

When a live snapshot contained both divergent `argsText` and a backend result, `ToolCallStreamControllerImpl.setResponse` closed the args stream first. That emitted `tool-call-args-text-finish`, so `ToolExecutionStream` parsed the stale incomplete prefix and resolved the reader with `Function parameter parsing failed` before the backend result chunk arrived.

Enqueueing the backend result before closing `argsText` preserves the authoritative response while retaining the existing close/flush behavior.

## Verification

- Regression red/green in `assistant-stream`:
  - old ordering: the focused test fails with `Function parameter parsing failed`
  - new ordering: the focused test receives `{ source: "backend" }`
- `assistant-stream`: 327 passed, 20 skipped
- `@assistant-ui/core`: 585 passed
- `aui-build` passed for both packages
- changed-file `oxlint`: 0 warnings, 0 errors
- changed-file `oxfmt --check`: passed
- `changeset status --since origin/main`: passed

The ordering regression lives in `assistant-stream` so Turbo tests do not depend on `@assistant-ui/core` resolving a concurrently built workspace package.

Closes #5102

### CODE REVIEW

**Review by Kinfe123 — CHANGES_REQUESTED** [review]
Thanks for the help. but got a real quick findings: 

1. The stream emits the correct backend result, but closing argsText still emits a second parse-error result afterward. The current test misses this because it drains the output without asserting its chunks.
2. Without a backend result , i think the original #5102 behavior remains,  divergent arguments close with the stale prefix and auto-submit an internal parse error through onResult. lets check and fix those gaps

**Comment by Solaris-star** [comment]
Addressed both findings in c95c44523:

- backend results now suppress the subsequent args parse/frontend execution path, so closing argsText cannot enqueue a second result
- the regression test now captures output chunks and asserts exactly one backend result while execute remains uncalled

Validation: assistant-stream 327 tests passed, package build passed, and oxfmt passed. PTAL.

**Review by Kinfe123 — APPROVED** [review]

File: packages/assistant-stream/src/core/modules/tool-call.ts [hunk 1]
```diff
@@ -65,9 +65,6 @@ class ToolCallStreamControllerImpl implements ToolCallStreamController {
   private _argsTextController!: TextStreamController;
 
   async setResponse(response: ToolResponseLike<ReadonlyJSONValue>) {
-    this._argsTextController.close();
-    await Promise.resolve(); // flush microtask queue
-    // TODO switch argsTextController to be something that doesn'#t require this
     this._controller.enqueue({
```
[the remark this answers is not in the document]
Comment (okisdev) [reply packages/assistant-stream/src/core/modules/tool-call.ts:68]: addressed in c95c44523: `ToolExecutionStream` now records tool call ids that received a backend result, and the args-text-finish handler skips the stale parse, the frontend execute, and the second result enqueue for those ids. the regression test pins this down by asserting `execute` is never called and exactly one result chunk reaches the output.

File: packages/assistant-stream/src/core/modules/tool-call.ts [hunk 2]
```diff
@@ -83,6 +80,9 @@ class ToolCallStreamControllerImpl implements ToolCallStreamController {
         ? { messages: response.messages }
         : {}),
     });
+    this._argsTextController.close();
```
[the remark this answers is not in the document]
Comment (okisdev) [reply packages/assistant-stream/src/core/modules/tool-call.ts:83]: addressed in c95c44523, deliberately via the consumer-side guard in `ToolExecutionStream` rather than a flag inside `setResponse`: the execution concern lives in the execution layer, and the guard also covers externally decoded streams, which on main can already deliver a result chunk before args-text-finish and would not be protected by a `setResponse`-internal flag. the regression test asserts `execute` stays uncalled and a single result chunk is emitted.

**Review by okisdev — APPROVED** [review]
## review

LGTM as `patch`. verified in a worktree: red/green on the ordering regression (reverting the `setResponse` reorder makes the test fail with exactly `Function parameter parsing failed` beating the backend result), and the full assistant-stream suite matches your numbers (327 passed, 20 skipped).

### shortcomings of the original implementation

- the initial commit reordered `setResponse` but left the subsequent `tool-call-args-text-finish` free to parse the stale prefix, run the frontend `execute`, and enqueue a second result; tools with side effects could run even though the backend had already answered.
- the regression test originally asserted only the reader response, so the double-result and double-execute paths were unpinned.

### what we modified

- (by the author, in c95c44523) `ToolExecutionStream` now records tool call ids that received a backend result; the finish handler skips the stale parse, the frontend execute, and the second enqueue for those ids, with cleanup in both `part-finish` branches. the test captures output chunks and asserts exactly one backend result with `execute` uncalled. as a side benefit the guard also protects externally decoded streams, which on main can already deliver `result` before `args-text-finish`.
- (repo side) the remaining #5102 scope moved to #5130 (divergent close without a backend result, snapshot adoption, auto-submit suppression, values-mode reconcile residual), so `Closes #5102` is accurate for this PR's slice.

### potential further fixes

- #5130 covers the shape this PR intentionally does not touch: with no backend result at close time (the #5098 interrupt scenario) a divergent close still auto-submits a parse-error result.
- the changeset also bumps `@assistant-ui/core` although the core diff is docs and a test comment only; a harmless no-op release, could be dropped.
- an optional test variant with complete valid args plus a backend result would pin that well-formed args also skip `execute`.
